cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RecordWriter.java

```diff
@@ -136,35 +185,122 @@ public StreamResult write(Iterator<Tuple2<DecoratedKey, Object[]>> sourceIterato
+
+        // token range for this partition is not among the write-replica-set ranges
+        if (!newRanges.contains(tokenRange))
+        {
+            Set<Range<BigInteger>> subRanges = getIntersectingSubRanges(newRanges, tokenRange);
+            // We have split ranges - likely resulting from pending nodes
+            if (subRanges.size() > 1)
+            {
+                // Create session using sub-range that contains the token from current row
+                Range<BigInteger> matchingRange = subRanges.stream().filter(r -> r.contains(token)).findFirst().get();
+                Preconditions.checkState(matchingRange != null,
+                                         String.format("Received Token %s outside of expected range %s", token, matchingRange));
+
+                if (streamSession != null && streamSession.getTokenRange() == matchingRange)
+                {
+                    return streamSession;
+                }
+                else
+                {
+                    LOGGER.debug(String.format("[{}] Creating stream session for range: %s", matchingRange), taskContext.partitionId());
+                    if (streamSession != null && batchSize != 0)
+                    {
+                        finalizeSSTable(streamSession, taskContext.partitionId(), sstableWriter, batchNumber, batchSize);
+                        sstableWriter = null;
+                        batchSize = 0;
+                    }
+                    return new StreamSession(writerContext, getStreamId(taskContext), matchingRange, failureHandler);
+                }
+            }
+        }
+
+        return (streamSession != null) ? streamSession : createStreamSession(taskContext);
+    }
+
+    /**
+     * Get ranges from the set that intersect and/or overlap with the provided token range
+     */
+    private Set<Range<BigInteger>> getIntersectingSubRanges(Set<Range<BigInteger>> ranges, Range<BigInteger> tokenRange)
+    {
+        return ranges.stream()
+                     .filter(r -> r.isConnected(tokenRange) && !r.intersection(tokenRange).isEmpty())
+                     .collect(Collectors.toSet());
+    }
+
+    private boolean haveTokenRangeMappingsChanged(TokenRangeMapping<RingInstance> startTaskMapping, TaskContext taskContext)
+    {
+        Range<BigInteger> taskTokenRange = getTokenRange(taskContext);
+        // Get the uncached, current view of the ring to compare with initial ring
+        TokenRangeMapping<RingInstance> endTaskMapping = writerContext.cluster().getTokenRangeMapping(false);
+        Map<Range<BigInteger>, List<RingInstance>> startMapping = taskTokenRangeMapping(startTaskMapping, taskTokenRange);
+        Map<Range<BigInteger>, List<RingInstance>> endMapping = taskTokenRangeMapping(endTaskMapping, taskTokenRange);
+
+        return !(startMapping.keySet().equals(endMapping.keySet()) &&
+               instancesFromMapping(startMapping).equals(instancesFromMapping(endMapping)));
```

Comment:
nit: I think this is easier to read with less parenthesis. 
```suggestion
        return !Objects.equals(startMapping.keySet(), endMapping.keySet()) ||
               !Objects.equals(instancesFromMapping(startMapping), instancesFromMapping(endMapping));
```

A follow up question. In the scenario that a node is joining at the beginning of a task and finishes joining at the end, the `haveTokenRangeMappingsChanged` would return false. Am I right?

Reply:
Thanks for the input.

We're checking if either the token-ranges or the set of instances differ between the start and end of the task to determine if it should fail the task (and retry).

In your example, if the the node is joining at the beginning of the task and has completed joining before the task ends, the above condition is still good as nothing would have changed and it would return `false`.

It is in the alternate case i.e. when the transitioning nodes appear during task execution that we want to detect and fail-fast.